### PR TITLE
Circuit Relay V2 client and DCUTR protocols for NAT Traversal

### LINKED ISSUE #533: NAT traversal implementation with Circuit Relay V2 and DCUtR protocols
This will be followed by #584 
Currently the demo uses 2 peers to connect to each other and connecting to central node in a demo environment(conferences, hotels) a NAT may be in place. 

Having this hole punching is available/relay node implementation could greatly improve availability.

### PR DESCRIPTION
**Pull request by MitaliBo**
## Description

Fixes #533 

This PR adds the behaviour of circuit relay v2 client and dcutr(direct connection upgrade through relay) which is needed for NAT Traversal either through relay or hole punching
## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).
- [x] I've requested a review  from `pyrsia/collaborators`.

### Code Contributions

- [x] I've built the code `cargo build --all-targets` successfully.
- [x] I've run the unit tests `cargo test --workspace` and everything passes.
- [x] I've made sure my rust toolchain is current `rustup update`.

### CODE REVIEW

File: src/network/client.rs [hunk 1]
```diff
@@ -124,6 +124,21 @@ impl Client {
...
+
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(Command::ListenRelay {
+                addr: addr.clone(),
+                sender,
+            })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not to be dropped.")
```
Comment (fishseabowl) [inline src/network/client.rs:139]: Please refer to PR #708.  It is better to propagate channel errors.

Comment (prince-chrismc) [reply src/network/client.rs:139]: 👍

File: pyrsia_node/src/args/parser.rs [hunk 2]
```diff
@@ -35,10 +36,35 @@ pub struct PyrsiaNodeArgs {
...
+    /// The mode (client-listen, client-dial) for dcutr or NoRelay
+    #[clap(long)]
+    pub mode: Mode,
+    /// The listening address
+    #[clap(long)]
+    pub relay_address: Option<Multiaddr>,
+}
+
+#[derive(Debug, Parser, PartialEq)]
+pub enum Mode {
```
Comment (prince-chrismc) [inline pyrsia_node/src/args/parser.rs:54]: This is the "Relay Mode" (not the node mode)

File: pyrsia_node/src/args/parser.rs [hunk 3]
```diff
@@ -35,10 +36,35 @@ pub struct PyrsiaNodeArgs {
...
+    Dial,
+    Listen,
+    NoRelay,
```
Comment (prince-chrismc) [inline pyrsia_node/src/args/parser.rs:57]: This is the old "direct connection mode"

File: src/network/client/command.rs [hunk 4]
```diff
@@ -32,6 +32,10 @@ pub enum Command {
         addr: Multiaddr,
         sender: oneshot::Sender<anyhow::Result<()>>,
     },
+    ListenRelay {
```
Comment (prince-chrismc) [inline src/network/client/command.rs:35]: This is the message routing component that needs to be handled

File: src/network/client/command.rs [hunk 5]
```diff
@@ -32,6 +32,10 @@ pub enum Command {
...
+        addr: Multiaddr,
+        sender: oneshot::Sender<Result<(), Box<dyn Error + Send>>>,
```
Comment (prince-chrismc) [inline src/network/client/command.rs:37]: Should be new anyhow as @fishseabowl mentioned above

**Review by prince-chrismc — COMMENTED** [review]
Some notes for my future self

**Review by prince-chrismc — COMMENTED** [review]
The main inspiration has been https://github.com/libp2p/rust-libp2p/tree/master/protocols/dcutr/examples

File: pyrsia_node/src/args/parser.rs [hunk 6]
```diff
@@ -35,10 +36,35 @@ pub struct PyrsiaNodeArgs {
...
     /// The maximum number of keys that can be provided on the network by this Pyrsia Node.
     #[clap(long, default_value = DEFAULT_MAX_PROVIDED_KEYS)]
     pub max_provided_keys: usize,
+    /// An address to connect with another Pyrsia Node (eg /ip4/127.0.0.1/tcp/45153/p2p/12D3KooWKsHbKbcVgyiRRgeXGCK4bp3MngnSU7ioeKTfQzd18B2v)
+    #[clap(long, short = 'P')]
+    pub peer: Option<Multiaddr>,
+    /// The mode (client-listen, client-dial) for dcutr or NoRelay
+    #[clap(long)]
+    pub mode: RelayMode,
+    /// The listening address
```
Comment (prince-chrismc) [inline pyrsia_node/src/args/parser.rs:48]:
```suggestion
    /// The address of the relay server. (local address for `Listen` or remote address for `Dial`
```

**Review by prince-chrismc — APPROVED** [review]

File: pyrsia_node/src/main.rs [hunk 7]
```diff
@@ -127,14 +129,32 @@ fn setup_http(args: &PyrsiaNodeArgs, p2p_client: Client) {
...
+            if let Some(relay_address) = args.relay_address {
+                handlers::dial_other_peer(p2p_client.clone(), &relay_address).await;
+            }
+        }
+        RelayMode::Listen => {
+            if let Some(relay_address) = args.relay_address {
+                p2p_client
+                    .listen_relay(&relay_address)
+                    .await
+                    .expect("Listening should not fail");
```
Comment (chb0github) [inline pyrsia_node/src/main.rs:143]: Dial should work the same

**Review by chb0github — CHANGES_REQUESTED** [review]
See comments

File: src/network/event_loop.rs [hunk 8]
```diff
@@ -301,6 +301,12 @@ impl PyrsiaEventLoop {
                     Err(e) => sender.send(Err(From::from(e))),
                 };
             }
+            Command::ListenRelay { addr, sender } => {
+                let _ = match self.swarm.listen_on(addr.with(Protocol::P2pCircuit)) {
```
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply src/network/event_loop.rs:305]: This is the convention being used the code today

image

**Review by chb0github — APPROVED** [review]
`let _` it be

**Comment by prince-chrismc** [comment]
> `let _` it be

After chatting on slack, the usage is pretty wide spread so we opted to make a new issue #756

**Pull request merged**